projects/plugins/jetpack/modules/widget-visibility/widget-conditions.php

```diff
@@ -699,25 +728,119 @@ static function generate_condition_key( $rule ) {
 	 * @return array Settings to display or bool false to hide.
 	 */
 	public static function filter_widget( $instance ) {
-		global $wp_query;
-
-		if ( empty( $instance['conditions'] ) || empty( $instance['conditions']['rules'] ) ) {
-			return $instance;
-		}
-
 		// Don't filter widgets from the REST API when it's called via the widgets admin page - otherwise they could get
 		// filtered out and become impossible to edit.
 		if ( strpos( wp_get_raw_referer(), '/wp-admin/widgets.php' ) && false !== strpos( $_SERVER['REQUEST_URI'], '/wp-json/' ) ) {
 			return $instance;
 		}
 
+		if ( ! empty( $instance['conditions']['rules'] ) ) {
+			// Legacy widgets: visibility found.
+			if ( self::filter_widget_check_conditions( $instance['conditions'] ) ) {
+				return $instance;
+			}
+			return false;
+		} elseif ( ! empty( $instance['content'] ) && has_blocks( $instance['content'] ) ) {
+			// Block-Based widgets: We have gutenberg blocks that could have the 'conditions' attribute
+			// Note: These blocks can be nested, and we would have to apply these conditions at any level.
+			$blocks = parse_blocks( $instance['content'] );
+			$blocks = self::recursively_filter_blocks( $blocks )[0];
+			if ( empty( $blocks ) ) {
+				return false;
+			}
+
+			$output = '';
+			foreach ( $blocks as $block ) {
+				if ( ! empty( $block ) ) {
+					$output .= render_block( $block );
+				}
+			}
+			$instance['content'] = $output;
+
+			return $instance;
+		}
+
+		// No visibility found.
+		return $instance;
+	}
+
+	/**
+	 * Recursively check the visibility conditions in blocks' attr.conditions.rules.
+	 * Any that fail the check will be removed.
+	 *
+	 * @param array $blocks (By reference; will be modified).
+	 * @return array [$blocks, $indexes_removed]
+	 */
+	public static function recursively_filter_blocks( &$blocks ) {
```

Comment:
Do you think it would be possible to add unit tests to these new methods?

Reply:
Added unit tests for this method here: 476ed773c17337cbfef55f721ef76e70e6dc4674
Notes:
* They are in the "wrong" directory. `./projects/plugins/jetpack/tests/php/modules/widgets/test_widget-visibility.php`. I wanted to place them in a new directory, ``./projects/plugins/jetpack/tests/php/modules/widget-visibility/`, but I couldn't get them to run inside that directory, even though `widget-visibility` is listed in the `projects/plugins/jetpack/tests/php/test-get-modules.php` file.
* In order to get the tests working, I had to: `jetpack docker sh`, then `cd /tmp/wordpress-develop/tests/phpunit/data`, then `svn up -r44722`, then `cd ../includes/` , then again `svn up -r44722`
  * After this `jetpack docker phpunit` would start running where it previously would not, but it would crash before completing all tests, and leave me without a report of failing tests. However, `jetpack docker phpunit -- --filter=widget` is sufficient to run my new test file without crashing.

Reply:
> then cd /tmp/wordpress-develop/tests/phpunit/data, then svn up -r44722, then cd ../includes/ , then again svn up -r44722

Reply:
> I wanted to place them in a new directory, ./projects/plugins/jetpack/tests/php/modules/widget-visibility/, but I couldn't get them to run inside that directory, even though widget-visibility is listed in the projects/plugins/jetpack/tests/php/test-get-modules.php file.

Try adding the new dir here:

https://github.com/Automattic/jetpack/blob/88ecfe04895c5014ec78f6558fa3030130314466/projects/plugins/jetpack/phpunit.xml.dist#L54-L56

That should do the trick!

Reply:
> Added unit tests for this method here: [476ed77](https://github.com/Automattic/jetpack/commit/476ed773c17337cbfef55f721ef76e70e6dc4674)

Thank you! ❤️